File: Assets/Scripts/Enemy/States/PatrolState.cs
using UnityEngine;
using ActionGameDemo.StateMachine;

namespace ActionGameDemo.Enemy
{
    /// <summary>
    /// 巡逻状态：在出生点周围随机移动，到达后等待。
    /// 检测到玩家后进入追击。
    /// </summary>
    public class PatrolState : EnemyState
    {
        private Vector3 _patrolDestination;
        private bool _isMoving;
        private float _waitTimer;

        public PatrolState(EnemyController owner, StateMachine<EnemyController> stateMachine)
            : base(owner, stateMachine) { }

        public override void EnterState()
        {
            _waitTimer = 0f;
            _isMoving = false;
            PickNewDestination();
        }

        public override void UpdateState()
        {
            if (_isMoving)
            {
                // 检查是否到达目的地
                if (owner.HasReachedPosition(_patrolDestination))
                {
                    _isMoving = false;
                    owner.StopMoving();
                    _waitTimer = 0f;
                }
            }
            else
            {
                _waitTimer += Time.deltaTime;
            }
        }

        public override void CheckSwitchState()
        {
            // 死亡检查
            if (Stats != null && Stats.CurrentHealth <= 0f)
            {
                SwitchTo(EnemyStateType.Dead);
                return;
            }

            // 检测到玩家 → 追击
            if (owner.DetectTarget())
            {
                SwitchTo(EnemyStateType.Chase);
                return;
            }

            // 低血量 → 逃跑
            if (owner.ShouldFlee() && owner.HasTarget)
            {
                SwitchTo(EnemyStateType.Flee);
                return;
            }

            // 等待超时 → 选择新巡逻点
            if (!_isMoving && _waitTimer >= owner.PatrolWaitTime)
            {
                PickNewDestination();
            }
        }

        public override void ExitState()
        {
        }

        private void PickNewDestination()
        {
            _patrolDestination = owner.GetRandomPatrolPoint();
            owner.MoveTo(_patrolDestination);
            _isMoving = true;
        }
    }
}

File: server/src/main/java/com/actiongame/server/ai/fsm/PatrolState.java
package com.actiongame.server.ai.fsm;

import com.actiongame.server.ai.decision.EnemyAIController;
import com.actiongame.server.util.Vector3;

/**
 * 巡逻状态 (对应Unity PatrolState)
 * 在出生点周围随机移动, 到达后等待. 检测到玩家→追击
 */
public class PatrolState extends EnemyState {
    private Vector3 patrolDestination;
    private boolean isMoving;
    private float waitTimer;

    public PatrolState(EnemyAIController controller) {
        super(controller);
    }

    @Override
    public void onEnter() {
        waitTimer = 0f;
        isMoving = false;
        pickNewDestination();
    }

    @Override
    public void onUpdate(float deltaTime) {
        if (isMoving) {
            if (controller.hasReached(patrolDestination)) {
                isMoving = false;
                controller.stopMoving();
                waitTimer = 0f;
            }
        } else {
            waitTimer += deltaTime;
        }
    }

    @Override
    public void onExit() {
    }

    public void checkSwitchState() {
        if (controller.isDead()) { switchTo(EnemyStateType.DEAD); return; }
        if (controller.detectTarget()) { switchTo(EnemyStateType.CHASE); return; }
        if (controller.shouldFlee() && controller.hasTarget()) { switchTo(EnemyStateType.FLEE); return; }
        if (!isMoving && waitTimer >= controller.getPatrolWaitTime()) { pickNewDestination(); }
    }

    private void pickNewDestination() {
        patrolDestination = controller.getRandomPatrolPoint();
        controller.moveTo(patrolDestination);
        isMoving = true;
    }
}
